### PR TITLE
Pull request support and enable supporting more webhooks

### PR DESCRIPTION
**Pull request by Shyri**
The main purppose of this pull request is make it easy to add support for more Bitbucket Webhooks than Repository push.

I've added the bases for an escalable plugin being able to progressively add support for other webhook events and actions.

I've added an hetero-list so one can set as many as webhooks triggers as he wants.

Finally I've added support for every kind of action over a Repository event (this means with no filter). Also for Pull Request events I've added support for any without filter and "approval" action.

I couldn't deal with the DSL support, i Think it's broken. The Trigger class now is DataboundConstructor which binds the hetero-list, and I don't know how it's done with DSL. Please could you review this part so I can fix it?

Thank you very much.

### CODE REVIEW

**Comment by fbelzunc** [comment]
Hello, I will review this very soon. Likely to be some time next week.

File: src/main/java/com/cloudbees/jenkins/plugins/BitBucketTrigger.java [hunk 1]
```diff
@@ -120,27 +107,27 @@ public void run() {
      * Returns the file that records the last/current polling activity.
      */
     public File getLogFile() {
-        return new File(job.getRootDir(),"bitbucket-polling.log");
+        return new File(job.getRootDir(), "bitbucket-polling.log");
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/BitBucketTrigger.java]: Unrelated change to the PR - even if it is wrong I prefer not to modify this and correct it on a separate PR

File: src/main/java/com/cloudbees/jenkins/plugins/BitBucketTrigger.java [hunk 2]
```diff
@@ -120,27 +107,27 @@ public void run() {
...
     }
 
     /**
      * Check if "bitbucket-polling.log" already exists to initialize it
      */
     public boolean IsLogFileInitialized() {
-        File file = new File(job.getRootDir(),"bitbucket-polling.log");
+        File file = new File(job.getRootDir(), "bitbucket-polling.log");
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/BitBucketTrigger.java]: Unrelated change to the PR - even if it is wrong I prefer not to modify this and correct it on a separate PR

File: src/main/java/com/cloudbees/jenkins/plugins/BitBucketTrigger.java [hunk 3]
```diff
@@ -120,27 +107,27 @@ public void run() {
...
         return file.exists();
     }
 
     @Override
     public DescriptorImpl getDescriptor() {
-        return (DescriptorImpl)super.getDescriptor();
+        return (DescriptorImpl) super.getDescriptor();
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/BitBucketTrigger.java]: Unrelated change to the PR - even if it is wrong I prefer not to modify this and correct it on a separate PR

File: src/main/java/com/cloudbees/jenkins/plugins/BitBucketTrigger.java [hunk 4]
```diff
@@ -120,27 +107,27 @@ public void run() {
...
     }
 
     /**
      * Action object for {@link Project}. Used to display the polling log.
      */
     public final class BitBucketWebHookPollingAction implements Action {
-        public Job<?,?> getOwner() {
+        public Job<?, ?> getOwner() {
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/BitBucketTrigger.java]: Unrelated change to the PR - even if it is wrong I prefer not to modify this and correct it on a separate PR

File: src/main/java/com/cloudbees/jenkins/plugins/BitBucketTrigger.java [hunk 5]
```diff
@@ -164,7 +151,7 @@ public String getLog() throws IOException {
          * Writes the annotated log to the given output.
          */
         public void writeLogTo(XMLOutput out) throws IOException {
-            new AnnotatedLargeText<BitBucketWebHookPollingAction>(getLogFile(), Charset.defaultCharset(),true,this).writeHtmlTo(0,out.asWriter());
+            new AnnotatedLargeText<BitBucketWebHookPollingAction>(getLogFile(), Charset.defaultCharset(), true, this).writeHtmlTo(0, out.asWriter());
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/BitBucketTrigger.java]: Unrelated change to the PR - even if it is wrong I prefer not to modify this and correct it on a separate PR

File: src/main/java/com/cloudbees/jenkins/plugins/BitBucketTrigger.java [hunk 6]
```diff
@@ -33,82 +33,69 @@
...
+            @Override
+            protected Job asJob() {
+                return job;
+            }
+        };
+
+        pJob.scheduleBuild2(5, new CauseAction(cause), bitbucketPayload);
+        if (pJob.scheduleBuild(cause)) {
+            String name = " #" + job.getNextBuildNumber();
+            LOGGER.info("SCM changes detected in " + job.getName() + ". Triggering " + name);
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/BitBucketTrigger.java]: Prefer:

```
LOGGER.log(Level.INFO, "SCM changes detected in {0}. Triggering {1}", new Object[] {job.getName(), name});
```

File: src/main/java/com/cloudbees/jenkins/plugins/BitBucketTrigger.java [hunk 7]
```diff
@@ -33,82 +33,69 @@
...
+        } else {
+            LOGGER.info("SCM changes detected in " + job.getName() + ". Job is already in the queue");
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/BitBucketTrigger.java]: ditto

File: src/main/java/com/cloudbees/jenkins/plugins/BitbucketEvent.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
+package com.cloudbees.jenkins.plugins;
+
+/**
+ * Created by isvillar on 11/03/2016.
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/BitbucketEvent.java]: Javadoc

File: src/main/java/com/cloudbees/jenkins/plugins/BitbucketEvent.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+ */
+public class BitbucketEvent {
+    public interface EVENT {
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/BitbucketEvent.java]: Javadoc

File: src/main/java/com/cloudbees/jenkins/plugins/BitbucketEvent.java [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+        String REPOSITORY = "repo";
+        String PULL_REQUEST = "pullrequest";
+    }
+
+    public interface REPOSITORY_ACTIONS {
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/BitbucketEvent.java]: Javadoc

File: src/main/java/com/cloudbees/jenkins/plugins/BitbucketEvent.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+        String PUSH = "push";
+    }
+
+    public interface PULL_REQUEST_ACTIONS {
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/BitbucketEvent.java]: Javadoc

File: src/main/java/com/cloudbees/jenkins/plugins/BitbucketEvent.java [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+        }
+
+        if(!checkAction(action)) {
+            throw new UnsupportedOperationException(action + " action for " + name + " event " );
+        } else {
+            this.action = action;
+        }
+    }
+
+    private boolean checkAction(String action) {
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/BitbucketEvent.java]: Javadoc

File: src/main/java/com/cloudbees/jenkins/plugins/BitbucketEvent.java [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+
+    public String getName() {
+        return name;
+    }
+
+    public String getAction() {
+        return action;
+    }
+
+    private boolean checkValidEvent(String name) {
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/BitbucketEvent.java]: Javadoc

File: src/main/java/com/cloudbees/jenkins/plugins/BitbucketPollingRunnable.java [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+import java.io.File;
+import java.io.IOException;
+import java.io.PrintStream;
+import java.text.DateFormat;
+import java.util.Date;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Created by Shyri Villar on 11/03/2016.
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/BitbucketPollingRunnable.java]: Javadoc

File: src/main/java/com/cloudbees/jenkins/plugins/cause/pullrequest/PullRequestCause.java [hunk 15]
```diff
@@ -0,0 +1,23 @@
+package com.cloudbees.jenkins.plugins.cause.pullrequest;
+
+import com.cloudbees.jenkins.plugins.cause.BitbucketTriggerCause;
+import com.cloudbees.jenkins.plugins.payload.BitbucketPayload;
+
+import java.io.File;
+import java.io.IOException;
+
+/**
+ * Created by isvillar on 15/03/2016.
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/cause/pullrequest/PullRequestCause.java]: Javadoc

File: src/main/java/com/cloudbees/jenkins/plugins/cause/repository/RepositoryCause.java [hunk 16]
```diff
@@ -0,0 +1,22 @@
+package com.cloudbees.jenkins.plugins.cause.repository;
+
+import com.cloudbees.jenkins.plugins.cause.BitbucketTriggerCause;
+import com.cloudbees.jenkins.plugins.payload.BitbucketPayload;
+
+import java.io.File;
+import java.io.IOException;
+
+/**
+ * Created by isvillar on 15/03/2016.
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/cause/repository/RepositoryCause.java]: Javadoc

File: src/main/java/com/cloudbees/jenkins/plugins/filter/BitbucketEventTriggerMatcher.java [hunk 17]
```diff
@@ -0,0 +1,10 @@
+package com.cloudbees.jenkins.plugins.filter;
+
+import com.cloudbees.jenkins.plugins.BitbucketEvent;
+
+/**
+ * Created by Shyri Villar on 14/03/2016.
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/filter/BitbucketEventTriggerMatcher.java]: Javadoc

File: src/main/java/com/cloudbees/jenkins/plugins/filter/BitbucketTriggerFilter.java [hunk 18]
```diff
@@ -0,0 +1,17 @@
...
+
+import com.cloudbees.jenkins.plugins.cause.BitbucketTriggerCause;
+import com.cloudbees.jenkins.plugins.payload.BitbucketPayload;
+import hudson.model.AbstractDescribableImpl;
+
+import java.io.File;
+import java.io.IOException;
+
+/**
+ * Created by Shyri Villar on 14/03/2016.
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/filter/BitbucketTriggerFilter.java]: Javadoc

File: src/main/java/com/cloudbees/jenkins/plugins/filter/BitbucketTriggerFilterDescriptor.java [hunk 19]
```diff
@@ -0,0 +1,8 @@
+package com.cloudbees.jenkins.plugins.filter;
+
+import hudson.model.Descriptor;
+
+/**
+ * Created by Shyri Villar on 14/03/2016.
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/filter/BitbucketTriggerFilterDescriptor.java]: Javadoc

File: src/main/java/com/cloudbees/jenkins/plugins/filter/FilterMatcher.java [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+import com.cloudbees.jenkins.plugins.BitbucketEvent;
+import com.cloudbees.jenkins.plugins.filter.pullrequest.PullRequestTriggerFilter;
+import com.cloudbees.jenkins.plugins.filter.pullrequest.PullRequestTriggerMatcher;
+import com.cloudbees.jenkins.plugins.filter.repository.*;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Created by Shyri Villar on 14/03/2016.
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/filter/FilterMatcher.java]: Javadoc

File: src/main/java/com/cloudbees/jenkins/plugins/filter/pullrequest/PullRequestActionDescriptor.java [hunk 21]
```diff
@@ -0,0 +1,9 @@
+package com.cloudbees.jenkins.plugins.filter.pullrequest;
+
+import hudson.model.Descriptor;
+
+/**
+ * Created by Shyri Villar on 14/03/2016.
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/filter/pullrequest/PullRequestActionDescriptor.java]: Javadoc

File: src/main/java/com/cloudbees/jenkins/plugins/filter/pullrequest/PullRequestAnyActionFilter.java [hunk 22]
```diff
@@ -0,0 +1,37 @@
...
+import com.cloudbees.jenkins.plugins.cause.pullrequest.PullRequestApprovedCause;
+import com.cloudbees.jenkins.plugins.payload.BitbucketPayload;
+import hudson.Extension;
+import org.kohsuke.stapler.DataBoundConstructor;
+
+import java.io.File;
+import java.io.IOException;
+
+/**
+ * Created by Shyri Villar on 14/03/2016.
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/filter/pullrequest/PullRequestAnyActionFilter.java]: Javadoc

File: src/main/java/com/cloudbees/jenkins/plugins/filter/pullrequest/PullRequestApprovedActionFilter.java [hunk 23]
```diff
@@ -0,0 +1,69 @@
...
+import net.sf.json.JSONArray;
+import net.sf.json.JSONObject;
+import org.kohsuke.stapler.DataBoundConstructor;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.logging.Logger;
+
+/**
+ * Created by isvillar on 14/03/2016.
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/filter/pullrequest/PullRequestApprovedActionFilter.java]: Javadoc

File: src/main/java/com/cloudbees/jenkins/plugins/filter/pullrequest/PullRequestCreatedActionFilter.java [hunk 24]
```diff
@@ -0,0 +1,34 @@
...
+import com.cloudbees.jenkins.plugins.cause.BitbucketTriggerCause;
+import com.cloudbees.jenkins.plugins.payload.BitbucketPayload;
+import hudson.Extension;
+import org.kohsuke.stapler.DataBoundConstructor;
+
+import java.io.File;
+import java.io.IOException;
+
+/**
+ * Created by Shyri Villar on 14/03/2016.
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/filter/pullrequest/PullRequestCreatedActionFilter.java]: Javadoc

File: src/main/java/com/cloudbees/jenkins/plugins/filter/pullrequest/PullRequestTriggerFilter.java [hunk 25]
```diff
@@ -0,0 +1,50 @@
...
+import hudson.Extension;
+import jenkins.model.Jenkins;
+import org.kohsuke.stapler.DataBoundConstructor;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * Created by Shyri Villar on 14/03/2016.
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/filter/pullrequest/PullRequestTriggerFilter.java]: Javadoc

File: src/main/java/com/cloudbees/jenkins/plugins/filter/pullrequest/PullRequestTriggerMatcher.java [hunk 26]
```diff
@@ -0,0 +1,24 @@
+package com.cloudbees.jenkins.plugins.filter.pullrequest;
+
+import com.cloudbees.jenkins.plugins.BitbucketEvent;
+import com.cloudbees.jenkins.plugins.filter.BitbucketEventTriggerMatcher;
+import com.cloudbees.jenkins.plugins.filter.BitbucketTriggerFilter;
+
+/**
+ * Created by Shyri Villar on 14/03/2016.
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/filter/pullrequest/PullRequestTriggerMatcher.java]: Javadoc

File: src/main/java/com/cloudbees/jenkins/plugins/filter/repository/RepositoryActionFilter.java [hunk 27]
```diff
@@ -0,0 +1,16 @@
...
+
+import com.cloudbees.jenkins.plugins.cause.BitbucketTriggerCause;
+import com.cloudbees.jenkins.plugins.payload.BitbucketPayload;
+import hudson.model.AbstractDescribableImpl;
+
+import java.io.File;
+import java.io.IOException;
+
+/**
+ * Created by Shyri Villar on 15/03/2016.
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/filter/repository/RepositoryActionFilter.java]: Javadoc

File: src/main/java/com/cloudbees/jenkins/plugins/filter/repository/RepositoryAnyActionFilter.java [hunk 28]
```diff
@@ -0,0 +1,34 @@
...
+import com.cloudbees.jenkins.plugins.cause.repository.RepositoryCause;
+import com.cloudbees.jenkins.plugins.payload.BitbucketPayload;
+import hudson.Extension;
+import org.kohsuke.stapler.DataBoundConstructor;
+
+import java.io.File;
+import java.io.IOException;
+
+/**
+ * Created by isvillar on 15/03/2016.
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/filter/repository/RepositoryAnyActionFilter.java]: Javadoc

File: src/main/java/com/cloudbees/jenkins/plugins/filter/repository/RepositoryPushActionFilter.java [hunk 29]
```diff
@@ -0,0 +1,35 @@
...
+import com.cloudbees.jenkins.plugins.cause.repository.RepositoryCause;
+import com.cloudbees.jenkins.plugins.payload.BitbucketPayload;
+import hudson.Extension;
+import org.kohsuke.stapler.DataBoundConstructor;
+
+import java.io.File;
+import java.io.IOException;
+
+/**
+ * Created by Shyri Villar on 15/03/2016.
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/filter/repository/RepositoryPushActionFilter.java]: Javadoc

File: src/main/java/com/cloudbees/jenkins/plugins/filter/repository/RepositoryTriggerFilter.java [hunk 30]
```diff
@@ -0,0 +1,50 @@
...
+import hudson.Extension;
+import jenkins.model.Jenkins;
+import org.kohsuke.stapler.DataBoundConstructor;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * Created by Shyri Villar on 11/03/2016.
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/filter/repository/RepositoryTriggerFilter.java]: Javadoc

File: src/main/java/com/cloudbees/jenkins/plugins/filter/repository/RepositoryTriggerMatcher.java [hunk 31]
```diff
@@ -0,0 +1,21 @@
+package com.cloudbees.jenkins.plugins.filter.repository;
+
+import com.cloudbees.jenkins.plugins.BitbucketEvent;
+import com.cloudbees.jenkins.plugins.filter.BitbucketEventTriggerMatcher;
+import com.cloudbees.jenkins.plugins.filter.BitbucketTriggerFilter;
+
+/**
+ * Created by Shyri villar on 15/03/2016.
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/filter/repository/RepositoryTriggerMatcher.java]: Javadoc

File: src/main/java/com/cloudbees/jenkins/plugins/payload/OldPostBitbucketPayload.java [hunk 32]
```diff
@@ -0,0 +1,21 @@
+package com.cloudbees.jenkins.plugins.payload;
+
+import net.sf.json.JSONObject;
+
+import javax.annotation.Nonnull;
+
+/**
+ * Created by isvillar on 15/03/2016.
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/payload/OldPostBitbucketPayload.java]: Javadoc

File: src/main/java/com/cloudbees/jenkins/plugins/payload/PullRequestPayload.java [hunk 33]
```diff
@@ -0,0 +1,30 @@
+package com.cloudbees.jenkins.plugins.payload;
+
+import hudson.EnvVars;
+import hudson.model.AbstractBuild;
+import net.sf.json.JSONObject;
+
+import javax.annotation.Nonnull;
+
+/**
+ * Created by Shryi Villar on 11/03/2016.
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/payload/PullRequestPayload.java]: Javadoc

File: src/main/java/com/cloudbees/jenkins/plugins/payload/RepositoryPayload.java [hunk 34]
```diff
@@ -0,0 +1,14 @@
+package com.cloudbees.jenkins.plugins.payload;
+
+import net.sf.json.JSONObject;
+
+import javax.annotation.Nonnull;
+
+/**
+ * Created by Shyri Villar on 15/03/2016.
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/payload/RepositoryPayload.java]: Javadoc

File: src/main/java/com/cloudbees/jenkins/plugins/processor/BitbucketPayloadProcessorFactory.java [hunk 35]
```diff
@@ -0,0 +1,33 @@
+package com.cloudbees.jenkins.plugins.processor;
+
+import com.cloudbees.jenkins.plugins.BitbucketEvent;
+import com.cloudbees.jenkins.plugins.BitbucketJobProbe;
+
+/**
+ * Created by isvillar on 11/03/2016.
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/processor/BitbucketPayloadProcessorFactory.java]: Javadoc

File: src/main/java/com/cloudbees/jenkins/plugins/processor/OldPostPayloadProcessor.java [hunk 36]
```diff
@@ -0,0 +1,27 @@
+package com.cloudbees.jenkins.plugins.processor;
+
+import com.cloudbees.jenkins.plugins.BitbucketEvent;
+import com.cloudbees.jenkins.plugins.BitbucketJobProbe;
+import com.cloudbees.jenkins.plugins.payload.BitbucketPayload;
+import com.cloudbees.jenkins.plugins.payload.OldPostBitbucketPayload;
+import net.sf.json.JSONObject;
+
+/**
+ * Created by Shyri Villar on 15/03/2016.
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/processor/OldPostPayloadProcessor.java]: Javadoc

File: src/main/java/com/cloudbees/jenkins/plugins/processor/OldPostPayloadProcessor.java [hunk 37]
```diff
@@ -0,0 +1,27 @@
...
+ */
+public class OldPostPayloadProcessor extends BitbucketPayloadProcessor {
+    public OldPostPayloadProcessor(BitbucketJobProbe jobProbe, BitbucketEvent bitbucketEvent) {
+        super(jobProbe, bitbucketEvent);
+    }
+
+    @Override
+    public void processPayload(JSONObject payload) {
+//        LOGGER.log(Level.INFO, "Received commit hook notification for {0}", repo);
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/processor/OldPostPayloadProcessor.java]: remove comment

File: src/main/java/com/cloudbees/jenkins/plugins/processor/PullRequestPayloadProcessor.java [hunk 38]
```diff
@@ -0,0 +1,27 @@
+package com.cloudbees.jenkins.plugins.processor;
+
+import com.cloudbees.jenkins.plugins.payload.BitbucketPayload;
+import com.cloudbees.jenkins.plugins.BitbucketEvent;
+import com.cloudbees.jenkins.plugins.BitbucketJobProbe;
+import com.cloudbees.jenkins.plugins.payload.PullRequestPayload;
+import net.sf.json.JSONObject;
+
+/**
+ * Created by Shyri Villar on 11/03/2016.
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/processor/PullRequestPayloadProcessor.java]: Javadoc

File: src/main/java/com/cloudbees/jenkins/plugins/processor/RepositoryPayloadProcessor.java [hunk 39]
```diff
@@ -0,0 +1,26 @@
+package com.cloudbees.jenkins.plugins.processor;
+
+import com.cloudbees.jenkins.plugins.BitbucketEvent;
+import com.cloudbees.jenkins.plugins.BitbucketJobProbe;
+import com.cloudbees.jenkins.plugins.payload.BitbucketPayload;
+import com.cloudbees.jenkins.plugins.payload.RepositoryPayload;
+import net.sf.json.JSONObject;
+
+/**
+ * Created by Shyri Villar on 11/03/2016.
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/processor/RepositoryPayloadProcessor.java]: Javadoc

File: src/test/java/com/cloudbees/jenkins/plugins/BitbucketEventTest.java [hunk 40]
```diff
@@ -0,0 +1,100 @@
...
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.runners.MockitoJUnitRunner;
+
+import javax.servlet.http.HttpServletRequest;
+
+import static org.junit.Assert.assertEquals;
+import static org.mockito.Mockito.when;
+/**
+ * Created by Shyri Villar on 11/03/2016.
```
Comment (fbelzunc) [inline src/test/java/com/cloudbees/jenkins/plugins/BitbucketEventTest.java]: Javadoc

File: src/test/java/com/cloudbees/jenkins/plugins/processor/PullRequestPayloadProcessorTest.java [hunk 41]
```diff
@@ -0,0 +1,99 @@
...
+import org.mockito.runners.MockitoJUnitRunner;
+
+import javax.servlet.http.HttpServletRequest;
+
+import static org.junit.Assert.assertEquals;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+/**
+ * Created by isvillar on 11/03/2016.
```
Comment (fbelzunc) [inline src/test/java/com/cloudbees/jenkins/plugins/processor/PullRequestPayloadProcessorTest.java]: Javadoc

File: src/main/java/com/cloudbees/jenkins/plugins/processor/BitbucketPayloadProcessor.java [hunk 42]
```diff
@@ -0,0 +1,20 @@
+package com.cloudbees.jenkins.plugins.processor;
+
+import com.cloudbees.jenkins.plugins.BitbucketEvent;
+import com.cloudbees.jenkins.plugins.BitbucketJobProbe;
+import net.sf.json.JSONObject;
+
+/**
+ * Created by Shyri Villar on 11/03/2016.
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/processor/BitbucketPayloadProcessor.java]: Javadoc

**Comment by fbelzunc** [comment]
@Shyri I am not sure if I am fine with the Pull Request support since there is a plugin for it - we recently rejected #25 because of this, so my question is:

* Why this part shouldn't be a part of the bitbucket-pullrequest-builder-plugin instead of here? Otherwise we will have duplicated stuff.

However, I like the approach of being able to trigger under several different actions .-)

**Comment by fbelzunc** [comment]
Also, I have the feeling that it will break a lot of set-ups because I didn't see the way a job with current *Build Trigger -> Trigger when ... to BitBucket* will be migrated to the current implementation. I have the feeling that this approach will perhaps stop triggering all those jobs until you don't manually re-configure them. Maybe I am wrong... if so, where did you manage this case?

**Comment by sarah-oakes** [comment]
@fbelzunc regarding the other bitbucket PR plugin and duplication of functionality; the other plugin polls bitbucket for changes rather than using webhooks to trigger builds. This functionality *is new* and more an extension of this plugins' current functionality rather than either extension or duplication of the other.

**Comment by Shyri** [comment]
Hi @fbelzunc 
Thank you very much for your revision. As @sarah-oakes states, the pull-request plugin only supports polling periodically in order to find new pull-requests and trigger jobs. My contribution in the other hand receives the webhook events from Bitbuckets as it did for push. And also sets the base for gradually supporting every webhook event as well as adding any other filter based on the webhooks payload content in a very easy way. As I did with "Trigger only when all reviewers have approved"

As the name of this plugin is "Bitbucket Plugin", it feels like it's a more generic 'official' plugin than the pull-request plugin. I thought it was a good idea to start creating an escalable way to support all the new webhooks in this plugin rather to have to install several plugins that doesn't meet all peoples's needs.

There are two things that may be broken, one is the DSL support and the other as @fbelzunc states the update for old version. I actually resolved it manually. I don't know how is this managed usually, is there any way to find previous configuration when updated or something?

**Comment by jrtheinert** [comment]
Hi all,

I have been exploring Jenkins plugins for my company for only about a week now.  We develop Android and iOS clients with a big server side development effort.  

This updated, more generic Bitbucket plugin is very close to what we are looking for to better automate our dev efforts on both client and server side.  Events vs. polling is very important for us.  IMO, the proposed version is not an update but a completely new plugin.  Maybe if we address it that way, as a new plugin, we can move this forward.  

I shared my thoughts on this with my teams and bosses and they see value in supporting this plugin.  To that end, I plan on tweaking Shyri’s original code to add some features we want.  My question then is how do I push those changes back to Shyri’s original repo?

Shyri, you did a very nice design/implementation job in your code.  That will make this easier to extend.

Thx

Jeff Theinert

**Comment by StFS** [comment]
I agree that this is a very much needed feature. Not being able to perform a build on PR creation is very limiting.

**Comment by StFS** [comment]
@jrtheinert I think you basically have two options:

1) Clone @Shyri 's repository (https://github.com/starzplayarabia/bitbucket-plugin), create a new branch off of his 'pull_request_support' branch, make your changes and commit them to your new branch and then make a pull request over to @Shyri 's repository. If you do that before this PR is merged he might be okay with merging those changes in and updating his PR here.

2) (this is probably a "better" way of doing this but might take longer, depending on whether @Shyri is willing to merge your PR before his PR gets merged here) Wait until this PR is merged into the official jenkinsci bitbucket-plugin repository and then do a separate PR with your enhancements.

File: src/main/java/com/cloudbees/jenkins/plugins/BitBucketTrigger.java [hunk 43]
```diff
@@ -1,114 +1,101 @@
...
+                                try {
+                                    cause = filter.getCause(getLogFile(), bitbucketPayload);
+
+                                    if (filter.shouldScheduleJob(bitbucketPayload)) {
+                                        scheduleJob(cause, bitbucketPayload);
+                                        return;
+                                    }
+
+                                } catch (IOException e) {
+                                    e.printStackTrace();
```
Comment (StFS) [inline src/main/java/com/cloudbees/jenkins/plugins/BitBucketTrigger.java]: Sorry to butt in here but I think this really should be using the `LOGGER` and the `e.printStackTrace(listener.error("Failed to schedule job"));` mechanism shouldn't it?

**Comment by Shyri** [comment]
@jrtheinert I am glad you like it. I am very happy to know you want to add some functionality. About creating a new plugin I don't know if we should create a new repository or JenkinsCI should instead. I think that if this is an official plugin named "Bitbucket Plugin", this functionality should be in here. But that is Jenkins people's call. 

For me it's ok if you want to fork our repo tweak it and make a pull request to us. I think it's going to take lees time for us to merge, and I suppose your contribution would automatically added to this PR. However if you want to wait until this PR is merged it's also ok for me. But it will take some more time I guess. Let's wait to see @fbelzunc says

File: src/main/java/com/cloudbees/jenkins/plugins/BitBucketTrigger.java (hunk 43 above)
Comment (Shyri) [inline src/main/java/com/cloudbees/jenkins/plugins/BitBucketTrigger.java]: @StFS I think you are right, I will change it and add it soon to the PR. Any other thing you think should be changed please let me know. 
Thanks!

**Comment by fbelzunc** [comment]
Hello all,

Your comments make sense and I have already received by e-mail some feedbacks with match your opinion here - you have convinced me. So I agree we can do this here.

I like a lot @Shyri ´s approach. 

> For me it's ok if you want to fork our repo tweak it and make a pull request to us. I think it's going to take lees time for us to merge, and I suppose your contribution would automatically added to this PR.

I think this makes sense.

I can work on the compatibility issue since it might be easier to me to figure out how to do it.

**Comment by fbelzunc** [comment]
Dear all,

I have created a [develop branch](https://github.com/jenkinsci/bitbucket-plugin/tree/develop) that we can use for this purpose.

@Shyri would you mind to launch your PR to develop branch instead? I will merge it and launch a PR from develop to master branch so we can better work all together.

Does it sound good?

**Comment by Shyri** [comment]
Ok @fbelzunc it sounds perfect for me.
I've created a new pull request #29 targeting develop branch. Also I've added a little change that I realized it was not pushed. Its a fix for checking only the users that are actually reviewers when setting the requirement of only trigger the job when all reviewers have approved.

Thank you very much

**Comment by fbelzunc** [comment]
Closing this PR as we are working in #30
